crml/rewards/src/lib.rs

```diff
@@ -0,0 +1,727 @@
+* limitations under the License.
+* You may obtain a copy of the License at the root of this project source code,
+* or at:
+*     https://centrality.ai/licenses/gplv3.txt
+*     https://centrality.ai/licenses/lgplv3.txt
+*/
+
+//! CENNZnet Staking Rewards
+//! This module handles the economic model for payouts of staking rewards for validators and their nominators.
+//! It also provides a simple treasury account suited for CENNZnet.
+//!
+//! The staking module should call into this module to trigger reward payouts at the end of an era.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use cennznet_primitives::{traits::StakerRewardPayment, types::Exposure};
+use frame_support::traits::OnUnbalanced;
+use frame_support::{
+	decl_event, decl_module, decl_storage,
+	traits::{Currency, Get, Imbalance},
+	weights::DispatchClass,
+};
+use frame_system::{self as system, ensure_root};
+use sp_arithmetic::{FixedI128, FixedPointNumber, FixedPointOperand};
+use sp_runtime::{
+	traits::{AccountIdConversion, One, Saturating, Zero},
+	ModuleId, Perbill,
+};
+use sp_std::{collections::vec_deque::VecDeque, prelude::*};
+
+/// The number of historical eras for which tx fee payout info should be retained.
+const HISTORICAL_PAYOUT_ERAS: usize = 7;
+
+/// A balance amount in the reward currency
+type BalanceOf<T> = <<T as Trait>::CurrencyToReward as Currency<<T as system::Trait>::AccountId>>::Balance;
+/// A pending increase to total issuance of the reward currency
+type PositiveImbalanceOf<T> =
+	<<T as Trait>::CurrencyToReward as Currency<<T as frame_system::Trait>::AccountId>>::PositiveImbalance;
+/// A pending decrease to total issuance of the reward currency
+type NegativeImbalanceOf<T> =
+	<<T as Trait>::CurrencyToReward as Currency<<T as frame_system::Trait>::AccountId>>::NegativeImbalance;
+
+pub trait Trait: frame_system::Trait {
+	/// The system event type
+	type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+	/// The reward currency system (total issuance, account balance, etc.) for payouts.
+	type CurrencyToReward: Currency<Self::AccountId>;
+	/// The treasury account for payouts
+	type TreasuryModuleId: Get<ModuleId>;
+}
+
+decl_event!(
+	pub enum Event<T>
+	where
+		Balance = BalanceOf<T>,
+	{
+		/// A reward payout happened (payout, remainder)
+		RewardPayout(Balance, Balance),
+	}
+);
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Rewards {
+		/// Inflation rate % to apply on reward payouts, it may be negative
+		pub InflationRate get(fn inflation_rate): FixedI128 = FixedI128::saturating_from_integer(1);
+		/// Development fund % take for reward payouts, parts-per-billion
+		pub DevelopmentFundTake get(fn development_fund_take) config(): Perbill;
+		/// Accumulated transaction fees for reward payout
+		pub TransactionFeePot get(fn transaction_fee_pot): BalanceOf<T>;
+		/// Historic accumulated transaction fees on reward payout
+		pub TransactionFeePotHistory get(fn transaction_fee_pot_history): VecDeque<BalanceOf<T>>;
```

Comment:
Is this going to grow infinitely?

Reply:
no, only the most recent will be kept in state